File: backend/src/controllers/UserController.js
const UserService = require('../services/UserService');

class UserController {
  constructor() {
    this.userService = new UserService();
  }

  async register(req, res) {
    try {
      const { name, email, password, role } = req.body;
      const user = await this.userService.register({ name, email, password, role });
      res.status(201).json(user);
    } catch (error) {
      res.status(400).json({ error: error.message });
    }
  }

  async login(req, res) {
    try {
      const { email, password } = req.body;
      const result = await this.userService.login(email, password);
      res.json(result);
    } catch (error) {
      res.status(401).json({ error: error.message });
    }
  }

  async getAllUsers(req, res) {
    try {
      const users = await this.userService.getAllUsers(req.user.role);
      res.json(users);
    } catch (error) {
      res.status(403).json({ error: error.message });
    }
  }

  async getUserById(req, res) {
    try {
      const user = await this.userService.getUserById(req.params.id, req.user);
      res.json(user);
    } catch (error) {
      if (error.message === 'Usuário não encontrado') {
        res.status(404).json({ error: error.message });
      } else {
        res.status(403).json({ error: error.message });
      }
    }
  }

  async deleteUser(req, res) {
    try {
     const { id } = req.params;
     const requestingUser = req.user;
     const result = await this.userService.deleteUser(id, requestingUser);
     res.status(200).json(result);
    } catch (error) {
      if (error.message === 'Usuário não encontrado') {
        res.status(404).json({ error: error.message });
      } else if (error.message.includes('acesso não autorizado')) {
        res.status(403).json({ error: error.message });
      } else {
        res.status(400).json({ error: error.message });
      }
    }
  }
}

module.exports = UserController;

File: backend/src/main/java/com/adminschool/backend/controller/UserController.java
package com.adminschool.backend.controller;

import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import com.adminschool.backend.dto.register.UserRegisterDTO;
import com.adminschool.backend.dto.UserResponseDTO;
import com.adminschool.backend.dto.login.LoginRequestDTO;
import com.adminschool.backend.dto.login.LoginResponseDTO;
import com.adminschool.backend.entity.User;
import com.adminschool.backend.security.JwtTokenProvider;
import com.adminschool.backend.services.UserService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/users")
@Validated
public class UserController {
    private final UserService userService;
    private final JwtTokenProvider jwtTokenProvider;

    public UserController(UserService userService, JwtTokenProvider jwtTokenProvider) {
        this.userService = userService;
        this.jwtTokenProvider = jwtTokenProvider;
    }

    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody @Validated UserRegisterDTO dto) {
        try {
            UserResponseDTO userResponse = userService.registerAndReturnDTO(dto);
            String token = jwtTokenProvider.generateToken(userService.authenticate(dto.email(), dto.password()).orElseThrow());
            return ResponseEntity.ok(Map.of("token", token, "user", userResponse));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("error", e.getMessage()));
        }
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody @Validated LoginRequestDTO loginRequest) {
        Optional<User> userOpt = userService.authenticate(loginRequest.login(), loginRequest.password());
        if (userOpt.isPresent()) {
            String token = jwtTokenProvider.generateToken(userOpt.get());
            LoginResponseDTO loginResponse = userService.authenticateAndReturnDTO(loginRequest.login(), loginRequest.password(), token).orElse(null);
            return ResponseEntity.ok(loginResponse);
        } else {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(Map.of("error", "Credenciais inválidas"));
        }
    }

    @GetMapping("/me")
    public ResponseEntity<?> getCurrentUser(@RequestParam String userId) {
        Optional<UserResponseDTO> userOpt = userService.findUserById(userId);
        if (userOpt.isPresent()) {
            return ResponseEntity.ok(userOpt.get());
        } else {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Usuário não encontrado"));
        }
    }
}
